### DISCUSSION TITLE
How to bring tree nodes into view from ViewModel side?

### DISCUSSION THREAD

**Discussion by DevAM-Tools** [AvaloniaUI/Avalonia#13475] [Q&A] [answered]
I have a big tree like structure that I want to show in a TreeView. Since the tree is quite big I load the nodes only on demand. (to minimize the memory consumption)

I want to allow the user to double click on a node to navigate to another node and bring it into view. Unfortunately the target node is not visible yet.

So my question is: Is there a way to bring tree nodes into view from the ViewModel side?

**Comment by timunie** [comment]
if you have a deep tree, I'd check TreeDataGrid as it has virtualization support. 

Anyway, my strategy would be to add a bool property to my NodeViewModel `IsExpanded` which is bound to TreeViewItem via a Style-Setter. Then you can expand the needed sub tree and after that select that given item.

**Comment by DevAM-Tools** [comment, marked as answer]
Thank you for the essential hint.

The following approach worked for me:

Bind `IsExpanded` Property via Style:
```
<Style Selector="TreeViewItem" x:DataType="vm:TreeNode">
    <Setter Property="IsExpanded" Value="{Binding IsExpanded}" />
</Style>
```

Enable `AutoScrollToSelectedItem` on the TreeView and bind `SelectedItem`:
```
<TreeView
    Name="MainTree"
    SelectedItem="{Binding SelectedNode}"
    ItemsSource="{Binding Nodes}"
    AutoScrollToSelectedItem="True">
</TreeView>
```

Then Expand all anchestor nodes of the target node from root to the target node. Finally set the selected Node to the target node.

```
Stack<TreeNode> anchestors = targetTreeNode.GetAncestors();
foreach (TreeNode anchestor in anchestors)
{
    anchestor.IsExpanded = true;
}

App.MainViewModel.SelectedNode = targetTreeNode;
```

Final note: I like the idea of TreeDataGrid but I don't think it is ready for production yet. I am experiencing many rendering issues together with "jumping" rows.

**Comment by DevAM-Tools** [comment]
Due to poor performance and massive memory consumption that comes along with huge trees I needed to rework my app and data structures accordingly to this example: https://github.com/AvaloniaUI/Avalonia/discussions/6365#discussioncomment-1137070
